### PR TITLE
MID-2846: Support for compression

### PR DESCRIPTION
**Pull request by irishshagua**
* Conditionally add compress Skipper filter to routes

Signed-off-by: bmooney <brian.mooney@zalando.ie>

### CODE REVIEW

**Comment by irishshagua** [comment]
I've only made it possible to configure the compression globally here as it's easy for clients to exclude compression by simply not proving an accept `Accept-Encoding: gzip` header. So I didn't see much of a point to add the route level overrides. Open to hearing other opinions on that though.

File: src/main/scala/ie/zalando/fabric/gateway/web/marshalling/JsonModels.scala [hunk 1]
```diff
@@ -178,6 +184,7 @@ trait JsonModels {
       whitelist                <- c.downField("x-fabric-whitelist").as[Option[Set[String]]]
       cors                     <- c.downField("x-fabric-cors-support").as[Option[CorsConfig]]
       employeeAccess           <- c.downField("x-fabric-employee-access").as[Option[EmployeeAccessConfig]]
+      compressionConfig        <- c.downField("x-fabric-compression-support").as[Option[CompressionConfig]]
```
Comment (cjbooms) [inline src/main/scala/ie/zalando/fabric/gateway/web/marshalling/JsonModels.scala:187]: Given compressionFactor 0 is disabled, can None be mapped to disabled instead?

Comment (irishshagua) [reply src/main/scala/ie/zalando/fabric/gateway/web/marshalling/JsonModels.scala:187]: No, even with compressionFactor 0, the response payload is effected

Comment (irishshagua) [reply src/main/scala/ie/zalando/fabric/gateway/web/marshalling/JsonModels.scala:187]: *Compression Factor 0*: 
```
curl -i -H "Authorization: Bearer $TOK" -H "Accept-Encoding: gzip" https://comp-test-gw.playground.zalan.do/resources --output -
HTTP/2 200
date: Fri, 04 Dec 2020 14:33:23 GMT
content-type: text/plain
content-encoding: gzip
server: Skipper
vary: Accept-Encoding

��R�Request served by gateway-test-app-6cc854df89-glg6s

HTTP/1.1 GET /resources

Host: comp-test-gw.playground.zalan.do
Accept-Encoding: gzip
Authorization: Bearer ...
Ot-Tracer-Spanid: 2e013ebc0ed49fb7
X-Forwarded-For: 86.41.171.11
X-Unverified-Audit: stups_fabric-event-scheduler
Ot-Tracer-Sampled: true
Ot-Tracer-Traceid: 26645183addee288
X-Amzn-Trace-Id: Root=1-5fca48b3-3191cf6e2512723059100396
User-Agent: curl/7.64.1
X-Flow-Id: iLcl-5XPZ4zdx1lG
X-Tokeninfo-Forward: {"realm":"/services","scope":["fabric-event-scheduler.write","uid"],"uid":"stups_fabric-event-scheduler"}
Accept: */*
X-Forwarded-Port: 443
X-Forwarded-Proto: https

�����q��%
```

*Compression Factor 4*:
```
curl -i -H "Authorization: Bearer $TOK" -H "Accept-Encoding: gzip" https://comp-test-gw.playground.zalan.do/resources --output -
HTTP/2 200
date: Fri, 04 Dec 2020 14:34:09 GMT
content-type: text/plain
content-encoding: gzip
server: Skipper
vary: Accept-Encoding

�Rws��]!#=xF�ê��ֲ���fW����������R��lkD�B�0�Ce^}�_�E���Z�m�#�`���l۲[Y�2Z*�����'z�f�!9�*
�D��u�#��`ОF��0�%u}{p�����������Q`0�Y�YR��2\���Gpt�k�T��X�r��(xޒ�����f��IΆbN����D�x!Y�PC1f�f������WC1rg��G�F�B���߆��D`-�;�<z�y�Mc�i^X$�->���Ux�"Q~a��}V'�s��<�[�/Ɓ,X�s�����3CtH�(t;D�������N�/��]<`����.�w�7/���j����h,;K�S_��UT�D�V�n#q�,�짵w�>�i{�e��~�Tl���|��Ż��N����凗�3N�<�@~D�F��眰qx�x��}����p���jVQx���^�w�?���]	��e<�u7K
�GKr���A�'�С�Y�<k�gh0�`0����R�����M��������k��滳��_�ch���
                                                          �8���1�w�+�jZ��t+�Km�AmϬู}�;k8�B~������*�%�7��AiM�\	� ���BB"�%XJ)L�`�ZY��O/�[+
                                                                                                                                          	,�
,�T��iJ	�-�L��V�{
��"�t�`|�n��g[�H�T��˧�uV�r֗�ۙ��@�8��n�#�&���0�2�U��-�1_��b�[IX�HZ%!���?)~+k�r���9[��T�
U3�4��WRL'�����B����@%��U�#�~���,�"C!�֛�d�^� -��;���^��su�2Q�H3�<oPY����_۵�k����x��oA�����o�������M�­%
```

*No CompressionFactor enabled*:
```
curl -i -H "Authorization: Bearer $TOK" -H "Accept-Encoding: gzip" https://comp-test-gw.playground.zalan.do/resources --output -
HTTP/2 200
date: Fri, 04 Dec 2020 14:34:55 GMT
content-type: text/plain
content-length: 1450
server: Skipper

Request served by gateway-test-app-6cc854df89-glg6s

HTTP/1.1 GET /resources

Host: comp-test-gw.playground.zalan.do
X-Tokeninfo-Forward: {"realm":"/services","scope":["fabric-event-scheduler.write","uid"],"uid":"stups_fabric-event-scheduler"}
User-Agent: curl/7.64.1
Accept-Encoding: gzip
Ot-Tracer-Sampled: true
Ot-Tracer-Spanid: 35b954c63b99aa2f
X-Forwarded-For: 86.41.171.11
X-Forwarded-Proto: https
Authorization: Bearer ...
Ot-Tracer-Traceid: 5374e5f13dea4
X-Amzn-Trace-Id: Root=1-5fca490f-08e769a94ec772b83e36c645
X-Flow-Id: 5Na7kEVn6BcuMcoJ
Accept: */*
X-Forwarded-Port: 443
X-Unverified-Audit: stups_fabric-event-scheduler
```

Comment (cjbooms) [reply src/main/scala/ie/zalando/fabric/gateway/web/marshalling/JsonModels.scala:187]: Strange one. I guess that's so client code doesn't have to do 
```
val payload = if (resposnse.isCompressed()) decompress(response.body) 
                       else response.body
```

**Comment by herojan** [comment]
:+1:

File: src/main/scala/ie/zalando/fabric/gateway/web/marshalling/JsonModels.scala (hunk 1 above)
Comment (irishshagua) [reply src/main/scala/ie/zalando/fabric/gateway/web/marshalling/JsonModels.scala:187]: Dunno. Would prob need to check with Skipper maintainers... 
I just didnt want to have any potential change from current workings...

**Comment by irishshagua** [comment]
👍

**Pull request merged**